changes BLAST parser code from perl to python

anilthanki opened the pull request:
and also adds functionality of filter by reciprocal results

anilthanki committed:
changes BLAST parser code from perl to python and also adds functionality of reciprocal results

anilthanki committed:
fixes travis error

anilthanki committed:
fixes more travis error

anilthanki committed:
fixes more travis error

nsoranzo commented on tools/blast_parser/blast_parser.py:
```diff
@@ -0,0 +1,52 @@
+            weight = 100
+
+            if line_cols[10] != "0" and line_cols[10] != "0.0":
+                weight = min(100, -1 * round(math.log10(float(line_cols[10])) / 2))
```
You can simplify these 2 lines as:
```python
            if float(line_cols[10]) != 0.0:
                weight = min(100, round(math.log10(line_cols[10]) / -2.0))
```

nsoranzo commented on tools/blast_parser/blast_parser.py:
```diff
@@ -0,0 +1,52 @@
+            if(options.reciprocal is False):
+                options.o.write("%s\t%s\t%d\n" % (line_cols[0], line_cols[1], weight))
+
+    if(options.reciprocal):
```
The parenthesis are not needed.

nsoranzo commented on tools/blast_parser/blast_parser.py:
```diff
@@ -0,0 +1,52 @@
+            else:
+                result[line_cols[0]] = {line_cols[1]: weight}
+
+            if(options.reciprocal is False):
```
```python
if not options.reciprocal:
```

nsoranzo commented on tools/blast_parser/blast_parser.py:
```diff
@@ -0,0 +1,52 @@
+
+    if(options.reciprocal):
+        for i in result:
+            pairs = result[i]
```
Replace these 2 lines with:
```python
        for i, pairs in result.items():
```

nsoranzo commented on tools/blast_parser/blast_parser.py:
```diff
@@ -0,0 +1,52 @@
+
+            if line_cols[0] in result:
+                if line_cols[1] not in result[line_cols[0]]:
+                    result[line_cols[0]].update({line_cols[1]: weight})
```
```python
                    result[line_cols[0]][line_cols[1]] = weight
```

nsoranzo commented on tools/blast_parser/blast_parser.py:
```diff
@@ -0,0 +1,52 @@
+                result[line_cols[0]] = {line_cols[1]: weight}
+
+            if(options.reciprocal is False):
+                options.o.write("%s\t%s\t%d\n" % (line_cols[0], line_cols[1], weight))
```
This write should be moved after the for loop in order to print only the first line for each pair of sequences (in case `-max_hsps 1` BLAST parameter was not used).

nsoranzo commented on tools/blast_parser/test-data/input.tabular:
```diff
@@ -31,7 +31,7 @@ ENSPFOT00000022544_poeciliaformosa_1	ENSXMAT00000001796_xiphophorusmaculatus_1	8
 ENSMICT00000002052_microcebusmurinus_1	ENSMICT00000002052_microcebusmurinus_1	100.00	92	0	0	1	92	1	92	1e-53	185
 ENSMICT00000002052_microcebusmurinus_1	ENSCAFT00000022963_canisfamiliaris_1	93.40	303	20	0	217	519	218	520	0.0	600
-ENSMICT00000002052_microcebusmurinus_1	ENSCAFT00000022963_canisfamiliaris_1	87.10	93	11	1	1	92	1	93	3e-46	164
+ENSMICT00000002052_microcebusmurinus_1	ENSCAFT00000022963_canisfamiliaris	87.10	93	11	1	1	92	1	93	3e-46	164
```
Not sure why you changed this?

nsoranzo committed:
Use an OrderedDict with sequence pairs as keys

- Clearer variable names
- Simplify weight formula
- Print only one line for each sequence pair

nsoranzo commented:
@anilthanki I've fixed the issues with the Python code, please double check my changes.

Only remaining doubt is the change to `tools/blast_parser/test-data/input.tabular` .

anilthanki commented:
You mean change i made in input.tabular, then it was only for testing if reciprocal works or not. We can revert it or keep it

nsoranzo committed:
Revert change to input.tabular

nsoranzo commented:
Yes, that change in input.tabular. I've reverted it now.
If my changes look good to you, feel free to merge.
